lambda/translcomprehension.ml

```diff
@@ -2,304 +2,406 @@ open Lambda
+        let to_binding =
+          binding Strict Pintval to_var (transl_exp ~scopes e3)
+        in
+        let low, high =
+          match dir with
+          | Upto -> Lvar from_var, Lvar to_var
+          | Downto -> Lvar to_var, Lvar from_var
+        in
+        let len_binding =
+          let init =
+            Lprim(Psubint, [Lprim(Paddint, [high; int 1], loc); low], loc)
+          in
+          binding Alias Pintval len_var init
+        in
+        let for_ = Lfor(id, Lvar from_var, Lvar to_var, dir, body) in
+        [from_binding; to_binding; len_binding], for_
+  in
+  bindings, len_var, for_
+
+(* Generate code to iterate over a comprehension block, along with some
+   initialising bindings.  The bindings will also bind the given
+   [length_var] ident to the total number2 of iterations in the
+   block. *)
+let iterate_arr_block ~transl_exp ~loc ~scopes
+      {clauses; guard} length_var body =
+  let body =
+    match guard with
+    | None -> body
+    | Some guard ->
+      Lifthenelse(transl_exp ~scopes guard, body, lambda_unit)
+  in
+  let body, lengths, rev_bindings =
+    List.fold_left
+      (fun (body, lengths, rev_bindings) clause ->
+         let new_bindings, new_length_var, body =
+           transl_arr_clause ~transl_exp ~scopes ~loc clause body
+         in
+         let rev_bindings = List.rev_append new_bindings rev_bindings in
+         let lengths = new_length_var :: lengths in
+         body, lengths, rev_bindings)
+      (body, [], []) clauses
+  in
+  let length_opt =
+    List.fold_left
+      (fun length var ->
+         match length with
+         | None -> Some (Lvar var)
+         | Some length -> Some (Lprim(Pmulint, [Lvar var; length], loc)))
+      None lengths
+  in
+  let length = Option.value length_opt ~default:(int 0) in
+  let length_binding = binding Alias Pintval length_var length in
+  let bindings = List.rev_append rev_bindings [length_binding] in
+  bindings, body
+
+let make_array_prim ~loc size init =
+  let prim =
     Primitive.simple ~name:"caml_make_vect" ~arity:2 ~alloc:true
   in
-  Lprim (Pccall prim_make_arr, [size; init], loc)
+  Lprim (Pccall prim, [size; init], loc)
 
-let make_array_of_kind kind size ~loc =
-  match kind with
-  | Pgenarray -> assert false (*An example is needed to create this array.*)
-  | Pintarray -> make_array size (int 0) ~loc
-  | Paddrarray -> make_array size (int 0) ~loc
-  | Pfloatarray ->
-    let prim_make_float_arr =
-      Primitive.simple ~name:"caml_make_float_vect" ~arity:1 ~alloc:true
-    in
-    Lprim (Pccall prim_make_float_arr, [size], loc)
+let make_floatarray_prim ~loc size =
+  let prim =
+    Primitive.simple ~name:"caml_make_float_vect" ~arity:1 ~alloc:true
+  in
+  Lprim (Pccall prim, [size], loc)
 
-let blit_array src src_pos dst dst_pos len ~loc=
+let blit_array_prim ~loc src src_pos dst dst_pos len =
   let prim_blit_arr =
     Primitive.simple ~name:"caml_array_blit" ~arity:5 ~alloc:true
```

Comment:
I just found out that there is a caml_floatarray_blit, which we might want to use.

Reply:
Good point. Looks like it would only save a single branch, so not a huge win, but probably still worth using. If we added a c function specialised to `Pintarray` that would actually be a much bigger win. I may try and do that later.